server/src/storage/mod.rs

```diff
@@ -0,0 +1,174 @@
+//! in-memory and in a parquet file data buffer, and storing compressed data.
+
+mod data_point;
+mod segment;
+
+use crate::storage::data_point::DataPoint;
+use crate::storage::segment::SegmentBuilder;
+use paho_mqtt::Message;
+use std::collections::vec_deque::VecDeque;
+use std::collections::HashMap;
+
+type Timestamp = i64;
+type Value = f32;
+type MetaData = Vec<String>;
+
+const INITIAL_BUILDER_CAPACITY: usize = 50;
+
+/// Keeping track of all uncompressed data, either in memory or in a file buffer. The data field should
+/// not be directly modified and is therefore only changed when using "insert_data".
+pub struct StorageEngine {
+    /// The uncompressed segments while they are being built.
+    data: HashMap<String, SegmentBuilder>,
+    /// Prioritized queue of finished segments that are ready for compression.
+    compression_queue: VecDeque<SegmentBuilder>,
+}
+
+impl StorageEngine {
+    pub fn new() -> Self {
+        StorageEngine {
+            // TODO: Maybe create with estimated capacity to avoid reallocation.
+            data: HashMap::new(),
+            compression_queue: VecDeque::new(),
+        }
+    }
+
+    /// Format `message` and insert it into the in-memory storage.
+    pub fn insert_message(&mut self, message: Message) {
+        match DataPoint::from_message(&message) {
+            Ok(data_point) => {
+                let key = data_point.generate_unique_key();
+
+                println!("Inserting data point {:?} into key '{}'.", data_point, key);
+
+                if let Some(segment) = self.data.get_mut(&*key) {
+                    println!("Found existing segment with key '{}'.", key);
+
+                    segment.insert_data(&data_point);
+
+                    if segment.is_full() {
+                        println!("Segment is full, moving it to the compression queue.");
+
+                        let finished_segment = self.data.remove(&*key).unwrap();
+                        self.compression_queue.push_back(finished_segment);
+                    }
+                } else {
+                    println!("Could not find segment with key '{}'. Creating segment.", key);
+
+                    let mut segment = SegmentBuilder::new(&data_point);
+                    segment.insert_data(&data_point);
+
+                    self.data.insert(key, segment);
+                }
+            }
+            Err(e) => eprintln!("Message could not be inserted into storage: {:?}", e),
+        }
+    }
+
+    /// If possible, return the oldest finished segment from the compression queue.
+    pub fn get_finished_segment(&mut self) -> Option<SegmentBuilder> {
+        self.compression_queue.pop_front()
+    }
+}
+
+#[cfg(test)]
+mod tests {
+    use super::*;
+    use rand::Rng;
+    use std::time::{SystemTime, UNIX_EPOCH};
+
+    fn insert_generated_message(storage_engine: &mut StorageEngine) -> String {
+        let value = rand::thread_rng().gen_range(0..100);
```

Comment:
Since the value has no effect on whether the tests fail it might be unnecessary to use proptest. Alternatively, a constant value could be used for the generated message instead of a random value.

Reply:
Then I suggest using a constant value instead of a random value to make it clear that the value has no effect. I thought the random value was used to specifically test multiple different value.

Reply:
Changed to use a constant value.